server/svix-server/src/queue/redis.rs

```diff
@@ -209,33 +375,98 @@ impl TaskQueueSend for RedisQueueProducer {
-            .arg(0u8);
-        let key: String = pool
-            .query_async(cmd)
+        let pool = self.pool.clone();
+        let main_queue_name = self.main_queue_name;
+        tokio::spawn(async move {
+            // TODO: Receive messages in batches so it's not always a Vec with one member
+            let mut pool = pool.get().await.unwrap();
+
+            // There is no way to make it await a message for unbounded times, so simply block for a short
+            // amount of time (to avoid locking) and loop if no messages were retreived
+            let resp = loop {
+                let resp: StreamReadReply = pool
+                    .query_async(Cmd::xread_options(
+                        &[main_queue_name],
+                        &[LISTEN_STREAM_ID],
+                        &StreamReadOptions::default()
+                            .group(WORKERS_GROUP, WORKER_CONSUMER)
+                            .count(1)
+                            .block(30_000),
+                    ))
+                    .await
+                    .unwrap();
+
+                if !resp.keys.is_empty() && !resp.keys[0].ids.is_empty() {
+                    break resp;
+                }
+            };
+
+            let element = &resp.keys[0].ids[0];
+            let id = element.id.clone();
+            let map = &element.map;
+
+            let task: QueueTask = if let Some(redis::Value::Data(data)) = map.get("data") {
+                serde_json::from_slice(data).expect("Invalid QueueTask")
+            } else {
+                panic!("No QueueTask associated with key");
+            };
+
+            tracing::trace!("RedisQueue: event recv <");
+
+            Ok(vec![TaskQueueDelivery { id, task }])
+        })
+        .await
+        .unwrap()
+    }
+}
+
+async fn migrate_v2_to_v3_queues(pool: &mut PooledConnection<'_>) {
+    migrate_list_to_stream(pool, LEGACY_V2_MAIN, MAIN).await;
+}
+
+async fn migrate_list_to_stream(pool: &mut PooledConnection<'_>, legacy_queue: &str, queue: &str) {
+    let batch_size = 1000;
+    loop {
+        let legacy_keys: Vec<String> = pool
+            .lpop(legacy_queue, NonZeroUsize::new(batch_size))
             .await
-            .map_err(|x| Error::Queue(x.to_string()))?;
-        tracing::trace!("RedisQueue: event recv <");
-        Ok(vec![from_redis_key(&key)])
+            .unwrap();
+        if legacy_keys.is_empty() {
+            break;
+        }
+        tracing::info!(
+            "Migrating {} keys from queue {}",
+            legacy_keys.len(),
+            legacy_queue
+        );
+
+        for key in legacy_keys {
+            let delivery = from_redis_key(&key);
+            let _: () = pool
+                .query_async(Cmd::xadd(
+                    queue,
+                    GENERATE_STREAM_ID,
+                    &[(QUEUE_KV_KEY, serde_json::to_string(&delivery.task).unwrap())],
+                ))
+                .await
+                .unwrap();
+        }
```

Comment:
Pipeline too, in batches of size X (whatever X is).

Pipeline is orders of magnitude faster, so it's especially important here.